### PR TITLE
[DISCUSS] Draft of Spec

### PR DESCRIPTION
**Pull request by philkra**
as discussed, this document provides us a platform to discuss:
* the fixed ordering of the k-v
* de-dot'ing and nesting of keys

Ref: https://github.com/elastic/ecs-logging-java/issues/51

### CODE REVIEW

File: spec/README.md
Comment (felixbarny) [inline spec/README.md]: `*` and `\` should also be replaced

File: spec/README.md
Comment (felixbarny) [inline spec/README.md]:
```suggestion
The substitution character should be `_`.
```

File: spec/spec.json
Comment (felixbarny) [inline spec/spec.json:5]: should we also include a version field in the log events themselves?

File: spec/spec.json
Comment (felixbarny) [inline spec/spec.json]:
```suggestion
            "required": true,
```

File: spec/spec.json
Comment (felixbarny) [inline spec/spec.json]:
```suggestion
            "required": true,
```

File: spec/spec.json
Comment (felixbarny) [inline spec/spec.json]:
```suggestion
            "required": true,
```

File: spec/README.md
Comment (felixbarny) [inline spec/README.md:22]: maybe mention that this should be in a single line (ND-JSON)

File: spec/README.md
Comment (codebrain) [reply spec/README.md]: Mention https://en.wikipedia.org/wiki/Snake_case?

File: spec/spec.json
Comment (philkra) [reply spec/spec.json:5]: I had the same idea, with appenders in mind! Will add that to the spec 💪

File: spec/README.md
Comment (philkra) [reply spec/README.md]: I'm currently looking for field naming constraints in the ES docs

File: spec/spec.json
Comment (webmat) [inline spec/spec.json]:
```suggestion
                    "substitute": "_"
```

File: spec/README.md
Comment (webmat) [inline spec/README.md]: I think this section should also touch on the discussion around dots in keys in general, as we've been discussing in https://github.com/elastic/ecs-logging-java/issues/51. I don't think we've reached a definitive decision yet, but perhaps we could start by documenting what was discussed today, as the proposal:

* Formatters should have exactly one dotted key -- `log.level` -- for readability, right after `@timestamp`, and the rest should be nested.
* If other issues or exceptions arise, this is where we would list them.
  * Would `trace.id` and `transaction.id` listed as well?
* Discuss how the dotted keys must be transformed to nesting ASAP, and how.

File: spec/README.md
Comment (webmat) [inline spec/README.md:18]: I like the idea of listing a few "levels" of viability.

I like the current MVP level of `@timestamp`, log.level, message & ecs.version. It should be possible to deliver that in most languages with minimal effort.

I think we should also describe a more fleshed out level, like the Java library lists [here](https://github.com/elastic/ecs-logging-java#mapping). I see this level as still mostly "drop-in" + perhaps configuration around the expensive attributes to collect. But essentially, developers are still logging like they used to, without needing to "care" about ECS field definitions.

We've discussed eventually supporting the possibility to let developers fill more ECS fields if they care. That could be one more level of maturity of a library. None of them is ready to do so yet (afaik), but we can still define this as possible level of maturity we will be working towards.

**Review by webmat — APPROVED** [review]
Thanks for starting this spec, this will be helpful in aligning the libraries.

Here's my comments so far. I don't mind if we merge aggressively and keep iterating on it as subsequent PRs, or getting that PR right before merging. Both approaches are fine for me. I'll let you decide how you want to approach this.

So after fixing the typo below, I'm LGTM :-)

**Comment by philkra** [comment]
@webmat , I added more elaborate examples to the README, but as you described correctly: we are not there yet with the "drop-ins". It should only illustrate what e.g. the . Java lib is already doing.

**Comment by webmat** [comment]
The ECS version field should not be dotted, it should be nested: `"ecs": {"version": "1.4.0"}`.

**Pull request merged**
